### PR TITLE
Add support for error cause in Neo4jError and Routing errors

### PR DESCRIPTION
**Pull request by bigmontz**
Root causes of routing issue were suppressed by the routing error message.

This change enables the `Neo4jError` object use the `cause` property for giving more context about the error. This
also improves the routing error by making use of the error `cause`.

`Error.cause` is a new feature in JS, see which runtime already implemented it:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/Error

### CODE REVIEW

File: packages/bolt-connection/src/connection-provider/connection-provider-routing.js [hunk 1]
```diff
@@ -473,19 +478,19 @@ export default class RoutingConnectionProvider extends PooledConnectionProvider
         }
 
         // try next router
-        const session = await this._createSessionForRediscovery(
+        const [session, error] = await this._createSessionForRediscovery(
           currentRouter,
           bookmarks,
           impersonatedUser
         )
         if (session) {
```
Comment (robsdedude) [inline packages/bolt-connection/src/connection-provider/connection-provider-routing.js:486]: Generally (not only here), the readability might benefit from short-circuiting along the lines of

```typescript
const [foo, error] = await foobar();
if (error) { // alternatively: `if (foo === null) {` or similar
  return [null, error]
}
// ...
```

File: packages/core/src/error.ts [hunk 2]
```diff
@@ -70,8 +70,9 @@ class Neo4jError extends Error {
    * @param {string} message - the error message
    * @param {string} code - Optional error code. Will be populated when error originates in the database.
    */
-  constructor (message: string, code: Neo4jErrorCode) {
-    super(message)
+  constructor (message: string, code: Neo4jErrorCode, cause?: Error) {
+    // @ts-expect-error
+    super(message, cause != null ? { cause } : undefined)
```
Comment (robsdedude) [inline packages/core/src/error.ts:75]: I am surprised this doesn't require a poly-fill. Why is this, do you have a link to docs or similar where I can educate myself?

File: packages/core/test/error.test.ts [hunk 3]
```diff
@@ -43,6 +51,31 @@ describe('newError', () => {
...
+    if (supportsCause) {
+      // @ts-expect-error
+      expect(error.cause).toBe(cause)
+    } else {
+      // @ts-expect-error
+      expect(error.cause).toBeUndefined()
+    }
+  })
+
+  test.each([null, undefined])('should create Neo4jErro without cause (%s)', (cause) => {
```
Comment (robsdedude) [inline packages/core/test/error.test.ts]:
```suggestion
  test.each([null, undefined])('should create Neo4jError without cause (%s)', (cause) => {
```

**Review by robsdedude — APPROVED** [review]

File: packages/bolt-connection/src/connection-provider/connection-provider-routing.js (hunk 1 above)
Comment (bigmontz) [reply packages/bolt-connection/src/connection-provider/connection-provider-routing.js:486]: Yes, but I'd like to change that much the code for this fix.

File: packages/core/src/error.ts (hunk 2 above)
Comment (bigmontz) [reply packages/core/src/error.ts:75]: The Error cause was added as an extra parameter to the Error constructor and Javascript doesn't enforce the number of parameters for calling a function. So, if i declare `function fn(a) { ... }` and you call as `fn('a', 'b')` without any issue. So, calling `Error('message', { cause: 'a' })` will not break any old implementation. 

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/Error

Comment (robsdedude) [reply packages/core/src/error.ts:75]: JS for you :upside_down_face: Thank's for the explanation.

**Pull request merged**